### ISSUE TITLE
Wrong links to dynamic zoom page in munin-cgi-html 

### ISSUE THREAD

**Issue by gjuric** [munin-monitoring/munin#460]
I understand that this option of generating graphs is deprecated but in the just released distribution of Debian (Jessie) Munin is in the latest stable version (2.0.25) which still uses this.

I have installed Munin, spawned fastgi processes for generating graphs and HTML and when I visit same pages the links to the dynamic zoom page is correct and looks like this:

```
/munin/static/dynazoom.html?cgiurl_graph=/...
```

But if I get to the same graph from the left navigation menu the link that is supposed to take me to the dynamic zoom page looks like this:

```
/munin/localdomain/localhost.localdomain/static/dynazoom.html?cgiurl_graph=/...
```

So of course, the second one does not work.

Is there an easy way to fix this?

**Comment by chteuchteu** [comment]
About your second link: were you on the `/munin/localdomain/localhost.localdomain/` page?
If so, there seems to be a missing leading `/` on the dynazoom link.

We'll try to find which link to edit: once done, you'll be able to edit the templates file (`.tmpl` files used for pages generation): once the HTML files will be regenerated, the links will be fixed :)

**Comment by gjuric** [comment]
Not sure anymore, I tried it again and this navigation path does not work:

    /munin/
    /munin/localdomain/localhost.localdomain/index.html#disk
    /munin/localdomain/localhost.localdomain/diskstats_latency/index.html
    /munin/localdomain/localhost.localdomain/diskstats_latency/sda.html
    /munin/localdomain/localhost.localdomain/diskstats_latency/static/dynazoom.html?cgiurl_graph=/cgi-bin/munin-cgi-graph&plugin_name=localdomain/localhost.localdomain/diskstats_latency/    sda&size_x=800&size_y=400&start_epoch=1430203806&stop_epoch=1430311806

This one works ok:

    /munin/
    /munin/localdomain/localhost.localdomain/index.html#network
    /munin/localdomain/localhost.localdomain/fw_conntrack.html
    /munin/static/dynazoom.html?cgiurl_graph=/cgi-bin/munin-cgi-graph&plugin_name=localdomain/localhost.localdomain/fw_conntrack&size_x=800&size_y=400&start_epoch=1430203806&stop_epoch=1430311806

I am using CGI strategy for generating graphs as I mentioned in the first post. Not sure if that has an impact.

**Comment by chteuchteu** [comment]
Could you try to edit the `munin-serviceview.tmpl` template in `/etc/munin/templates`?
There are 4 links that should be edited. They all look like `<a href="<TMPL_VAR NAME="ZOOMDAY">">`, and should be (for your configuration) `<a href="/munin/<TMPL_VAR NAME="ZOOMDAY">">`. (notice the leading `/munin/` in the href attribute).

This should do the trick: `static/dynazoom.html[...]` will become `/munin/static/dynazoom.html`

**Comment by gjuric** [comment]
There is something fishy going on. One page works, and links look like this:

    <td><a href="../../static/dynazoom...

Then I kill the FastCGI process and start it again and then the same page doesn't work because links then look like this:

    <td><a href="static/dynazoom....

I can not find a pattern which page is going to have wrong links when I restart the FastCGI process.

**Comment by chteuchteu** [comment]
Looks weird indeed.
I don't know how to help you here, maybe @steveschnepp will

**Comment by gjuric** [comment]
In case it helps, this is the nginx config I am using:

    location /munin/static/ {
        alias /etc/munin/static/;
    }

    location /munin/ {
        fastcgi_split_path_info ^(/munin)(.*);
        fastcgi_param PATH_INFO $fastcgi_path_info;
        fastcgi_pass unix:/var/run/munin/fcgi-html.sock;
        include fastcgi_params;
    }

    location ^~ /cgi-bin/munin-cgi-graph/ {
        fastcgi_split_path_info ^(/cgi-bin/munin-cgi-graph)(.*);
        fastcgi_param PATH_INFO $fastcgi_path_info;
        fastcgi_pass unix:/var/run/munin/fcgi-graph.sock;
        include fastcgi_params;
    }

Also, I have these settings in munin.conf:

    graph_strategy cgi
    html_strategy cgi
    cgiurl_graph /cgi-bin/munin-cgi-graph

I appreciate all the help I can get.

**Comment by matteocontrini** [comment]
Same configuration, same issue

**Comment by Phoenix1o1** [comment]
I gotta say, I heve this problem too. I've been looking around on the web, and i tried different solution suggested, but im still stuck with a broken zoom. Any help appreciated. To be clear, its a fresh install, Lamp server and all, and literally nothing else installed apart from Munin

**Comment by dim-0** [comment]
I also can confirm this issue.
Seems to be connected to the CGI job rendering the HTML.
If you let cron do the HTML and use CGI for the graphs only, it works (on my server).
The CGI seems to have some issues with the link **source** location (instead of the **target**). In some places I saw the HTML document always refering to "./static/<whatever>", no matter where I've been.

**Comment by gjuric** [comment]
I see that the issue is closed. Which version of Munin fixed this? Is there a simple patch/diff we can apply to the version installed from Debian stable?

**Comment by chteuchteu** [comment]
Hi @gjuric ,
This issue isn't closed: I referenced it from a [chteuchteu/Munin-for-Android](https://github.com/chteuchteu/Munin-for-Android) issue which is now closed ;)

**Comment by gjuric** [comment]
You are right, the UI confused me :+1:

**Comment by MoonSweep** [comment]
I Just stumbled on this bug, too. On my Munin server, it happens only if I access a graph from the domain page (but not from an individual host's page) : for some reason, munin-cgi-html inserts the domain name between "https://myserver.tld/munin" and "/static/dynazoom.html?...". Editing the URL in the address bar by simply removing this wrongfully added domain name makes the dynamic zoom page appear.

**Comment by MatthewVance** [comment]
I'm experiencing this issue also. I'm using Munin version 2.0.25-1 on Raspbian Jessie Lite with NGINX as the web server. I'm also using cgi (I think html method works).

In my case, it does not appear to be inserting the domain name between `https://myserver.tld/munin` and `/static/dynazoom.html?...` but rather `host-group-name/host-name`. If I remove that portion of the URL, the dynamic zoom page loads correctly. 

From what I can tell, the dynamic zoom related links don't work if I try to access those when starting from the navigation links on the left menu. However, the links work, at least in some cases, if I start from the home page and select one of the links in the main part of the page.

The relevant portion of my NGINX config is:

```
  location / {
    rewrite ^/$ munin/ redirect; break;
  }

  location /munin/ {
    fastcgi_split_path_info ^(/munin)(.*);
    fastcgi_param PATH_INFO $fastcgi_path_info;
    fastcgi_pass unix:/var/run/munin/fcgi-html.sock;
    include fastcgi_params;

    # auth_basic            "Restricted";
    # Create the htpasswd file with the htpasswd or openssl passwd tool.
    # auth_basic_user_file  /etc/nginx/htpasswd;
    # expires modified +310s;
  }

  location ^~ /munin-cgi/munin-cgi-graph/ {
    access_log off;
    fastcgi_split_path_info ^(/munin-cgi/munin-cgi-graph)(.*);
    fastcgi_param PATH_INFO $fastcgi_path_info;
    fastcgi_pass unix:/var/run/munin/fcgi-graph.sock;
    include fastcgi_params;
  }

  location /munin/static/ {
    alias /etc/munin/static/;
    # expires modified +1w;
  }
```

**Comment by AlexanderSk** [comment]
Hello, perhaps this article may help you. I once followed the instructions for a Centos 5 Dedi and i made it work with zoom! [http://www.journal.deviantdev.com/munin-dynazoom-suexec/](url)

**Comment by gjuric** [comment]
Thanks, but your URL is not "clickable" and the article does not have anything to do with this issue as far as I can tell.

**Comment by matteocontrini** [comment]
The link works but the article talks about Apache

**Comment by eamdog** [comment]
This nginx rule rewrites the broken links and makes them point to the right graphs again.

`rewrite ^(.*)/group\.name/hostname.*(/static/.*)$ $1$2;`

**Comment by nicomen** [comment]
Same issue on Ubuntu 14.04 with Munin 2.0.19-3 - but it seems so random. Suddenly some of the dynazoom links in some checks are wrong.

**Comment by miminashi** [comment]
Same issue on Raspbian jessie with 2.0.33 (it comes from debian backports).

https://github.com/munin-monitoring/munin/issues/460#issuecomment-97536936
It works for me as a workaround.

**Comment by miminashi** [comment]
Same issue on Raspbian jessie with Munin 2.0.33 (comes from jessie backports).
Absolutely same as https://github.com/munin-monitoring/munin/issues/460#issuecomment-237002811

**Comment by donoles** [comment]
Hi guys. Same issue on Ubuntu 16.04, with 2.0.25. It runs with fcgid. And it's not random. When I go to 'system-day.html' first, and then click the some cpu page, for instance, then the links are wrong. I do apachectl restart, or simply kill the munin-cgi-html process spawned by apache, and reload the cpu.html page, and links are OK again. I bet that's because some global variables remain from previous requests, in the same munin-cgi-html process, and make it behave differently.

**Comment by brumela** [comment]
Same issue on Debian 9 (munin 2.0.33-1, nginx 1.10.3-1+deb9u1, spawn-fcgi v1.6.4).
When it stops working I execute 
`/etc/init.d/munin-fastcgi restart`
and then works again.

**Comment by nathanwoodruffoc** [comment]
Debian 9.5, munin 2.0.33-1, apache 2.4.25
I put this redirect rule in my /etc/munin/apache24.conf file as a workaround and it does the job:
RedirectMatch ^/munin-cgi/munin-cgi-html/(.+)/static/dynazoom.html(.*) /munin-cgi/munin-cgi-html/static/dynazoom.html$2

**Comment by 0xE02B7** [comment]
Also have this problem:

**NGINX**
nginx version: nginx/1.12.2

**MUNIN**
munin-nginx-2.0.43-1.el7.noarch
munin-cgi-2.0.43-1.el7.noarch
munin-node-2.0.43-1.el7.noarch
munin-common-2.0.43-1.el7.noarch
munin-2.0.43-1.el7.noarch

**OS**
CentOS Linux release 7.6.1810 (Core)
3.10.0-862.6.3.el7.x86_64

Not working link (browsed via sidepanel).
I removed the beginnig and end of the URI.
**\***/munin/$GROUPNAME/$SERVICEGROUP/$FQDN/memcached_multi_evictions/static/dynazoom.html?cgiurl_graph=/**$1**

Working link:
**\***/munin/static/dynazoom.html?cgiurl_graph=/**$1**

Is there any chance this issue can be fixed soon?

**Comment by gjuric** [comment]
I doubt it. It has been 4 years since I've reported this bug.

**Comment by sumpfralle** [comment]
I never managed to reproduce this behavior on any of the munin servers I maintain. Thus it seems to be a tricky beast ...

Specifically I fail to follow the directions regarding links from the side-panel. Maybe someone of you could describe exactly which links should be followed up the point where a wrong link is visible on the following public munin instance: https://monitor.opennet-initiative.de/ ?
(it runs with apache2 and fcgid-based graphs)

**Comment by gjuric** [comment]
I have a feeling that people with this issue are running it behind nginx.

**Comment by sumpfralle** [comment]
> I have a feeling that people with this issue are running it behind nginx.

At least some of the reporters do that, yes.
The URL I linked above was previously delivered by nginx, too. We changed it to apache for unrelated reasons, thus I have no nginx-based public munin instance to offer for your review.

Maybe another approach: could someone please share the full html source of the page (and its URL), that contains the broken reference?

**Comment by gnuheidix** [comment]
@sumpfralle as you wish :-)

https://gist.github.com/gnuheidix/fcaf6168f65b88a0592d42d0944f4be1

**Comment by sumpfralle** [comment]
@gnuheidix: thank you!

I am surprised by the output.
Could you please locate the file `HTMLConfig.pm` in your system and share the output of the following command with us?
```shell
grep static/dynazoom SOME_PATH/Munin/Master/HTMLConfig.pm
```

Thank you!

**Comment by gnuheidix** [comment]
@sumpfralle: you're very welcome, thanks for your time on this issue.

requested investigation:
```
# grep -ni static/dynazoom /usr/share/perl5/vendor_perl/Munin/Master/HTMLConfig.pm
505:	my $common_url = "$root_path/static/dynazoom.html?cgiurl_graph=$config->{'cgiurl_graph'}&plugin_name=$path&size_x=$size_x&size_y=$size_y";
```

Package as provided from Arch Linux ARM repositories: http://mirror.archlinuxarm.org/armv7h/extra/munin-2.0.26-5-any.pkg.tar.xz

SHA512SUM of this package:
```
09c74acba1db3259ffc203f8f11a5e18e1f9850cd8b51ec34b82ad2978e3299212f26b43c2e1a051ef22f589bce7646bf9b14e0891023a44623195c7b79c1725  munin-2.0.26-5-any.pkg.tar.xz
```

**Comment by sumpfralle** [comment]
Thank you for the details!

The mystery seems to grow without bounds :(

The problematic HTML source you provided contains the following line:
```html
<td><a href="static/dynazoom.html?cgiurl_graph=...
```

The corresponding line in the template (`master/www/munin-serviceview.tmpl`) is the following:
```html
<td><a href="<TMPL_VAR NAME="ZOOMDAY">">
```

The above `zoomday` is defined like this:
```
my $common_url = "$root_path/static/dynazoom.html?cgiurl_graph=...
$srv{zoomday} = "$common_url&amp;start_epoch=$start_day&amp;stop_epoch=$epoch_now";
```

I tried to look for a possible explanation for the output being `...href="static/...`. But the above code snippet clearly and unconditionally prepends `static/dynazoom` with a slash. I cannot see a way how this leading slash could get lost on its way to the browser.

The only possible explanation for this situation seems to be some kind of post-processing. I am quite certain that this is not happening within munin. Given the provided nginx configuration for the case above (see https://gist.github.com/gnuheidix/fcaf6168f65b88a0592d42d0944f4be1), I also tend to think, that no such substitution is done by nginx (or the fastcgi module).
Or maybe I am missing some subtle behavior in any of these components?

@gnuheidix: Maybe you could try to manipulate the `$common_url` in `Munin/Master/HTMLConfig.pm` and check how this influences the output? Maybe this could shed some light on this mystery.

**Comment by gnuheidix** [comment]
@sumpfralle jup, this is a tricky one. :-)

I've played around with `$common_url` and noticed that changes will only make it to the CGI-rendered HTML after munin-cron ran. I've identified that munin-cron puts some information into _/var/lib/munin/htmlconf.storable_ It seems that everything is correct in there. The content of this file doesn't change until the next munin-cron execution.
```
# grep -a dynazoom /var/lib/munin/htmlconf.storable | cut -d? -f1 | sort | uniq
�../../static/dynazoom.html
�../../static/dynazoom.html
�../../static/dynazoom.html
�../../static/dynazoom.html
�../../static/dynazoom.html
�../../static/dynazoom.html
�../../static/dynazoom.html
�../../static/dynazoom.html
�../../static/dynazoom.html
�../../static/dynazoom.html
�../../static/dynazoom.html
�../../static/dynazoom.html
�../../static/dynazoom.html
�../../static/dynazoom.html
�../../static/dynazoom.html
�../../static/dynazoom.html
�../../static/dynazoom.html
�../../static/dynazoom.html
�../../static/dynazoom.html
�../../static/dynazoom.html
```
Furthermore, I've noticed that after the munin-html-cgi service gets restarted or the munin-cron gets executed, the HTML rendering works correctly for a short period of time. After more testing, I found an easy way to break the HTML rendering by issuing certain requests one after another.

As soon as a category overview page gets requested, the HTML rendering of subsequent category detail graphs breaks.

Details:
``` 
$ curl --silent http://hannelore/daheim/hannelore/lpstat.html | grep dynazoom | cut -d? -f1
  <td><a href="../../static/dynazoom.html
  <td><a href="../../static/dynazoom.html
  <td><a href="../../static/dynazoom.html
  <td><a href="../../static/dynazoom.html
$ curl --silent http://hannelore/daheim/hannelore/lpstat.html | grep dynazoom | cut -d? -f1
  <td><a href="../../static/dynazoom.html
  <td><a href="../../static/dynazoom.html
  <td><a href="../../static/dynazoom.html
  <td><a href="../../static/dynazoom.html
$ curl --silent --output /dev/null http://hannelore/printing-day.html 
$ curl --silent http://hannelore/daheim/hannelore/lpstat.html | grep dynazoom | cut -d? -f1
  <td><a href="static/dynazoom.html
  <td><a href="static/dynazoom.html
  <td><a href="static/dynazoom.html
  <td><a href="static/dynazoom.html
$ curl --silent http://hannelore/daheim/hannelore/lpstat.html | grep dynazoom | cut -d? -f1
  <td><a href="static/dynazoom.html
  <td><a href="static/dynazoom.html
  <td><a href="static/dynazoom.html
  <td><a href="static/dynazoom.html
$ sudo systemctl restart munin-html
$ curl --silent http://hannelore/daheim/hannelore/lpstat.html | grep dynazoom | cut -d? -f1
  <td><a href="../../static/dynazoom.html
  <td><a href="../../static/dynazoom.html
  <td><a href="../../static/dynazoom.html
  <td><a href="../../static/dynazoom.html
```
I assume that there is a global state hidden somewhere within the munin-html-cgi process which has a negative effect as soon as it gets asked for a category overview page.

**Comment by gnuheidix** [comment]
One more thing:
I was able to eliminate nginx from the problem space. I've used the program cgi-fcgi provided by [fcgi](https://archlinuxarm.org/packages/armv7h/fcgi) to issue requests directly to the munin-html-cgi process.

The following script is based on the tests I made through nginx which I've documented in my last comment. I've replicated most of the FastCGI parameters which are also being issued by nginx. I've identified them by using the [debug_connection](https://nginx.org/en/docs/ngx_core_module.html#debug_connection) directive.

```bash
#!/usr/bin/env bash
PATH_INFO=/daheim/hannelore/lpstat.html \
QUERY_STRING= \
REQUEST_METHOD=GET \
SCRIPT_NAME=/munin \
REQUEST_URI=/daheim/hannelore/lpstat.html \
DOCUMENT_URI=/munin/daheim/hannelore/lpstat.html \
GATEWAY_INTERFACE=CGI/1.1 \
cgi-fcgi -bind -connect /run/munin/fcgi-html.sock | grep dynazoom | cut -d? -f1

PATH_INFO=/printing-day.html \
QUERY_STRING= \
REQUEST_METHOD=GET \
SCRIPT_NAME=/munin \
REQUEST_URI=/printing-day.html \
DOCUMENT_URI=/munin/printing-day.html \
GATEWAY_INTERFACE=CGI/1.1 \
cgi-fcgi -bind -connect /run/munin/fcgi-html.sock > /dev/null && echo "Bang!"

PATH_INFO=/daheim/hannelore/lpstat.html \
QUERY_STRING= \
REQUEST_METHOD=GET \
SCRIPT_NAME=/munin \
REQUEST_URI=/daheim/hannelore/lpstat.html \
DOCUMENT_URI=/munin/daheim/hannelore/lpstat.html \
GATEWAY_INTERFACE=CGI/1.1 \
cgi-fcgi -bind -connect /run/munin/fcgi-html.sock | grep dynazoom | cut -d? -f1

systemctl restart munin-html && echo "Restarted"

PATH_INFO=/daheim/hannelore/lpstat.html \
QUERY_STRING= \
REQUEST_METHOD=GET \
SCRIPT_NAME=/munin \
REQUEST_URI=/daheim/hannelore/lpstat.html \
DOCUMENT_URI=/munin/daheim/hannelore/lpstat.html \
GATEWAY_INTERFACE=CGI/1.1 \
cgi-fcgi -bind -connect /run/munin/fcgi-html.sock | grep dynazoom | cut -d? -f1
```
Script output:
```
  <td><a href="../../static/dynazoom.html
  <td><a href="../../static/dynazoom.html
  <td><a href="../../static/dynazoom.html
  <td><a href="../../static/dynazoom.html
Bang!
  <td><a href="static/dynazoom.html
  <td><a href="static/dynazoom.html
  <td><a href="static/dynazoom.html
  <td><a href="static/dynazoom.html
Restarted
  <td><a href="../../static/dynazoom.html
  <td><a href="../../static/dynazoom.html
  <td><a href="../../static/dynazoom.html
  <td><a href="../../static/dynazoom.html
```

**Comment by sumpfralle** [comment]
@gnuheidix: what a beautiful and minimal recipe! Thanks a lot!

Based on your procedure I finally managed to reproduce the problem. The problematic changes were introduced in f285904b9956d952b1b06de185805cce265740b2 and bcd60dd752e8ebe85b317cd3a0932fc4c11078af. These commits are a slightly hacky workaround for different paths to be used when referencing the `dynazoom` resource in different contexts.

I just prepared a pull request that should fix this issue: #1163.

I would be glad, if someone could verify that my fix (2fde4da9a17dd2273bed94cb51a1ed0ff60f802b) really solves the issue.
